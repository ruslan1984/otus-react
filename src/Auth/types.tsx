type auth = {
  user: string;
  password: string;
  authorized: boolean;
};

export enum CheckState {
  initiated,
  succeed,
  failed,
}
