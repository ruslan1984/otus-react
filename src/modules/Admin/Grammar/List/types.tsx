export type GrammarList = Array<GrammarListItem>;
export type GrammarListItem = {
  id: number;
  name: string;
};
export type GrammarListType = {
  list: GrammarList;
  loading: boolean;
};
