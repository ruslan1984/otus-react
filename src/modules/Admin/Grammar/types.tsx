export type grammar = {
  id: number;
  name: string;
  text: string;
};
export type grammarList = {
  id: number;
  name: string;
};
