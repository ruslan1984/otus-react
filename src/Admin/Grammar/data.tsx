let data = [
    {
        id: 1,
        name: "Страница 1",
        text: "Контент 1",
    },
    {
        id: 2,
        name: "Страница 2",
        text: "Контент 2",
    },
];

export function grammarList() {
    const result = data.map((item) => {
        return {
            id: item.id,
            name: item.name,
        };
    });

    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve(result);
        }, 1000);
    });
}

export const grammarDetail = (id: number) => {
    let result: any;
    data.forEach((item) => {
        if (id == item.id) {
            result = {
                name: item.name,
                text: item.text,
            };
        }
    });

    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve(result);
        }, 1000);
    });
};
export function updateDetail(id: number, update_data: any) {
    data = data.map((item) => {
        if (id == item.id) {
            return {
                id: item.id,
                name: update_data.name,
                text: update_data.text,
            };
        }
        return item;
    });
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve(true);
        }, 1000);
    });
}
