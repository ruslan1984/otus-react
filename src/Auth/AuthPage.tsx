import React, { FC, useState, useEffect, useCallback } from "react";
import Presenter from "./Presenter";
import { Redirect } from "react-router-dom";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { actions } from "@store/reducers/auth/auth";
import { CheckState } from "@auth/types";

interface AuthProps {
    user: string;
    password: string;
    login: any;
    loading: boolean;
    status: any;
    setUser: any;
    setPassword: any;
}
const AuthPage: FC<AuthProps> = (props: AuthProps) => {
    const [user, setUser] = useState("");
    const [password, setPassword] = useState("");
    const submit = async (e: React.FormEvent) => {
        e.preventDefault();
        props.login({ user: user, password: password });
    };

    if (props.status === CheckState.succeed) {
        return <Redirect to="/admin" />;
    }
    return (
        <Presenter
            user={user}
            password={password}
            loading={props.loading}
            changeUser={(ev: React.FormEvent) =>
                setUser((ev.target as HTMLInputElement).value)
            }
            changePassword={(ev: React.FormEvent) =>
                setPassword((ev.target as HTMLInputElement).value)
            }
            submit={submit}
        />
    );
};
const mapStateToProps = (state: any) => {
    return {
        status: state.auth.status,
        loading: state.auth.loading,
    };
};
const mapDispatchToProps = (dispatch) => {
    return {
        login: bindActionCreators(actions.login, dispatch),
    };
};
export default connect(mapStateToProps, mapDispatchToProps)(AuthPage);
