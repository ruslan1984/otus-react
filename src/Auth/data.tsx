import { sleep } from "@/functions";
export let data = {
    user: "root",
    password: "root",
};
export const login = async (user: string, password: string) => {
    await sleep(200);

    return user == data.user && password == data.password;
};

export const logout = async () => {
    await sleep(200);
    localStorage.removeItem("auth");
};

export const isAuthorised = async () => {
    await sleep(200);
    return localStorage.getItem("auth") === data.user;
};
