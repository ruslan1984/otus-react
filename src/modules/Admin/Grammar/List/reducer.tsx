import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { GrammarListType, GrammarList, Loading } from "./types";

export const defaultState: GrammarListType = {
  list: [],
  loading: Loading.ok,
};
export const grammarListSlice = createSlice({
  name: "grammarList",
  initialState: defaultState,
  reducers: {
    setList: (state) => state,
    fetch: (state, { payload }: PayloadAction<GrammarList>) => {
      return { ...state, list: payload };
    },
    setLoading: (state, { payload }: PayloadAction<Loading>) => {
      return { ...state, loading: payload };
    },
  },
});
export const { actions, reducer } = grammarListSlice;
