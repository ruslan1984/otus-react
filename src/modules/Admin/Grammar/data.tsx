import * as R from "ramda";

import { grammar } from "./types";
import { sleep } from "@/functions";

const data: Array<grammar> = [
  {
    id: 1,
    name: "Страница 1",
    text: "Контент 1",
  },
  {
    id: 2,
    name: "Страница 2",
    text: "Контент 2",
  },
];

export async function grammarList() {
  await sleep(1000);
  return R.project(["id", "name"], data);
}

export const grammarDetail = async (id: number) => {
  await sleep(1000);
  return data.find((item) => item.id == id);
};

export async function updateDetail(id: number, updateData: grammar) {
  await sleep(1000);
  const update: grammar | undefined = data.find((item) => item.id == id);
  if (!update) return false;
  update.name = updateData.name;
  update.text = updateData.text;
  return true;
}
