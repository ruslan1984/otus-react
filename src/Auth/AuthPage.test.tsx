import React from "react";
import { mount } from "enzyme";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import reducer from "@store/reducers";
import { call, fork } from "redux-saga/effects";
import createSagaMiddleware from "redux-saga";
import { BrowserRouter } from "react-router-dom";
import AuthPage from "@auth/AuthPage";
import { loginSaga,checkUserSession, login, logout } from "@store/reducers/auth/sagas";
import { sleep } from '@/functions';
import { data, login as dataLogin } from '@auth/data';
import {
    isAuthorised
} from "@auth/session";
import { CheckState } from "@auth/types";

const sagaMiddleware = createSagaMiddleware();

function* rootSaga() {
    yield fork(loginSaga);
}

// if (process.env.NODE_ENV === `development`) {
//     middleware.push(loginMiddleware);
// }
export const store = configureStore({
    reducer,
    middleware: [sagaMiddleware],
});
sagaMiddleware.run(rootSaga);


describe( "Auth saga", () => {
    it("checkUserSession", async ()=>{
        const generator = checkUserSession();
        expect(generator.next().value).toEqual(call(isAuthorised));
        expect(generator.next().value).toMatchSnapshot();
        expect(generator.next().done).toBe(true)
    });
    it("login ok", async () => {

        const generator = login( {
            payload:{user:data.user,password:data.password}
            }
        );
        expect(generator.next().value.payload.action).toEqual( { type: 'auth/loading', payload: true });
        expect(generator.next().value).toEqual(call(dataLogin,data.user,data.password));
        generator.next(true);
        expect(generator.next().value.payload.action).toEqual( { type: 'auth/auth', payload: 1 });
        expect(generator.next().value.payload.action).toEqual( { type: 'auth/loading', payload: false });
        expect(generator.next().done).toBe(true);
    });
    it("login not ok", async () => {
        const generator = login( {
            payload:{user:data.user+'1',password:data.password}
            }
        );
        expect(generator.next().value.payload.action).toEqual( { type: 'auth/loading', payload: true });
        expect(generator.next().value).toEqual(call(dataLogin,data.user+'1',data.password));
        generator.next(false);
        expect(generator.next().done).toBe(true);
    });
    it("logout", async ()=>{
        const generator = logout();
        expect(generator.next().value).toMatchSnapshot();
        expect(generator.next().value.payload.action).toEqual( { type: 'auth/auth', payload: 0 });
        expect(generator.next().done).toBe(true);
    });

});

describe( "Auth form", () => {

    const wrapper = mount(
        <Provider store={ store }>
            <BrowserRouter>
                <AuthPage />
            </BrowserRouter>
        </Provider>
    );
    const name = wrapper.find( "input[name='name']" );
    const password = wrapper.find( "input[name='password']" );
    const button = wrapper.find( "button" );

    wrapper.find( 'input' ).at( 0 ).simulate( 'change', { target: { name: 'name', value: data.user } } );

    it( "should not login #login", async () => {
        wrapper.find( 'input' ).at( 1 ).simulate( 'change', { target: { name: 'password', value: data.password + '1' } } );
        wrapper.find( 'form' ).simulate( "submit", { preventDefault: () => null } );
        await sleep( 1000 );
        expect( store.getState().auth.status ).toEqual( CheckState.initiated );
    } );
    
    it( "should login #login", async () => {
        wrapper.find( 'input' ).at( 1 ).simulate( 'change', { target: { name: 'password', value: data.password } } );
        wrapper.find( 'form' ).simulate( "submit", { preventDefault: () => null } );
        await sleep( 1000 );
        expect( store.getState().auth.status ).toEqual( CheckState.succeed );
    } );
} );