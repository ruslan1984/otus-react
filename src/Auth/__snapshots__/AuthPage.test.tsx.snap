// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth saga checkUserSession 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": 0,
      "type": "auth/auth",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`Auth saga logout 1`] = `Promise {}`;
